hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3393,8 +3393,12 @@ private TableGroup consumeEntityJoin(SqmEntityJoin<?> sqmJoin, TableGroup lhsTab
 				null,
 				() -> p -> predicate.set( combinePredicates( predicate.get(), p ) ),
 				this
-				);
+		);
 		getFromClauseIndex().register( sqmJoin, tableGroup );
+		// For joins, we also need to register the table groups for the treats
```

Comment:
Might be worth extracting that to a method, because we would have to do the same for normal roots and cross joins, as well as do something similar for attribute joins.

Reply:
We already did this for attribute joins (I copied this from there), but we did not for from clause roots and cross joins. I extracted this to a method and added it for all cases you mentioned.